src/MediatR/Mediator.cs

```diff
@@ -41,20 +41,34 @@ public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, Cancellation
         }
 
         public Task Publish<TNotification>(TNotification notification, CancellationToken cancellationToken = default)
-            where TNotification : INotification
+             where TNotification : INotification
         {
             if (notification == null)
             {
                 throw new ArgumentNullException(nameof(notification));
             }
-
-            var notificationType = notification.GetType();
+            
+            var notificationType = typeof(TNotification);

```

Comment:
this should really be GetType() like it was, as TNotification can be an interface/abstract type. Especially in this case, since in the added Publish method, the notification is checked and casted to INotification, so TNotification == INotification and it wouldn't find any handlers.

Note this isn't covered by any tests, but has been brought up before.

Reply:
I don't think the generic version can use the overload. The generic version needs to use `typeof(TInterface)` and the non-generic one needs `notification.GetType()`

It's a separate issue to default to GetType - I'd rather not.

Reply:
See https://github.com/jbogard/MediatR/issues/124

It needs the concrete type, not the type of T decided by the compiler.

e.g.

```cs
var events = new IEvent[] { new EventA(), new EventB(), new EventC() };
events.Select(e => _mediatr.Publish(e));
```

TNotification is IEvent there, that's not going to work.